File: p441e/arrange_coins.cpp
// g++ -std=c++11 *.cpp -o test && ./test && rm -f test
#include <cmath>
#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    int arrangeCoins(int n) {
        long l = 0, r = n / 2 + 1;
        while (l < r) {
            long m = (l + r + 1) / 2;
            long sum = m * (m + 1);
            if (sum <= 2 * static_cast<long>(n))
                l = m;
            else
                r = m - 1;
        }
        return static_cast<int>(l);
    }
    int arrangeCoinsMath(int n) {
        return (int)((sqrt(1 + 8 * (long)n) - 1) / 2);
    }
};
// TEST
struct Test {
    int n;
    int expected;
    void run() {
        Solution sol;
        int actual = sol.arrangeCoins(n);
        cout << "The number of full staircase rows of " 
             << n << " -> " << actual << endl;
        assert(actual == expected);
        actual = sol.arrangeCoinsMath(n);
        assert(actual == expected);
    }
};

int main(int argc, char* argv[]) {
    vector<Test> tests = {
        {5, 2},
        {8, 3},
        {9, 3},
        {10, 4}
    };
    for (auto &t: tests) t.run();
    return 0;
}

File: p441e/arrange_coins.py
class Solution:
    def arrangeCoins(self, n: int) -> int:
        lo, hi = 0, n // 2 + 1
        while lo < hi:
            m = (lo + hi) // 2 + 1
            if m * (m + 1) <= 2 * n:
                lo = m
            else:
                hi = m - 1
        return lo


# TESTS
for n, expected in [
    (0, 0),
    (1, 1),
    (2, 1),
    (3, 2),
    (5, 2),
    (8, 3),
]:
    sol = Solution()
    actual = sol.arrangeCoins(n)
    print("Given n =", n, "the total # of full staircase rows ->", actual)
    assert actual == expected
